### PR TITLE
Add `nvmath-python` & `mathdx`

### PR DESCRIPTION
**Pull request by leofang**
This PR adds two NVIDIA packages to conda-forge:
1. `nvmath-python`: A Python library for NVIDIA Math Libraries
2. `MathDx`: A header-only, C++ template library bundling NVIDIA cuFFTDx and cuBLASDx 

cc: @conda-forge/cuda for vis

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by leofang** [comment]
This is ready. Ping:
- @conda-forge/help-python-c for reviewing `nvmath-python`
- @conda-forge/help-c-cpp for reviewing `mathdx`
- @conda-forge/cuda for vis

FYI:
> * Detected pure Python wheel(s) in source: `https://files.pythonhosted.org/packages/py3/n/nvidia-mathdx/nvidia_mathdx-24.4.0-py3-none-any.whl`. This is discouraged. Please consider using a source distribution (sdist) instead.

this is expected because NVIDIA policy is to only release wheels and not sdist (we reserve sdist for other purposes), plus for this particular release (24.04.0) MathDx does not release in any other forms (ex: tarballs). In a future release, we'll switch to use the tarball once available.

**Comment by leofang** [comment]
There seems to be non-deterministic overlinking errors... Are there known bugs in conda-build?

File: recipes/nvmath-python/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    sha256: 151ce8f50b9e65082efc1edaf0aa01b5a96650f7226ae18a3b47af286dfd62fb
+
+build:
+  number: {{ build }}
+  skip: true  # [py < 39 or osx or ppc64le]
+
+outputs:
+
+  - name: nvmath-python
+    version: {{ ".".join(version.split(".")[:3]) }}
```
Comment (jakirkham) [inline recipes/nvmath-python/meta.yaml:32]:
```suggestion
    version: {{ version }}
```

File: recipes/nvmath-python/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        - cython <3                              # [build_platform != target_platform]
+      host:
+        - python
+        - pip
+        - cython <3
+        # We use headers from CUDA 12 for simplicity, but they can come from any version.
+        # On conda-forge using the CUDA 12 packages is more lightweight.
+        - cuda-version 12.*                      # [cuda_compiler_version == "None"]
+        - cuda-cudart-dev                        # [cuda_compiler_version == "None"]
+        - cuda-crt                               # [cuda_compiler_version == "None"]
```
Comment (jakirkham) [inline recipes/nvmath-python/meta.yaml:72]:
```suggestion
        - cuda-crt
```

File: recipes/nvmath-python/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        - export CUDA_PATH=$BUILD_PREFIX/targets/x86_64-linux/   # [linux64]
+        - export CUDA_PATH=$BUILD_PREFIX/targets/sbsa-linux/     # [aarch64]
+        - set CUDA_PATH=%PREFIX%\Library                         # [win64]
+        {% elif cuda_major == 11 %}
+        # do nothing, CUDA_PATH is supposed to be set in the CI image
+        {% elif cuda_major == 0 %}
+        - export CUDA_PATH=$PREFIX/targets/x86_64-linux/         # [linux64]
+        - export CUDA_PATH=$PREFIX/targets/sbsa-linux/           # [aarch64]
+        - set CUDA_PATH=%PREFIX%\Library                         # [win64]
+        {% endif %}
```
Comment (jakirkham) [inline recipes/nvmath-python/meta.yaml:49]:
```suggestion
        - export CUDA_PATH=$PREFIX/targets/x86_64-linux/         # [linux64]
        - export CUDA_PATH=$PREFIX/targets/sbsa-linux/           # [aarch64]
        - set CUDA_PATH=%PREFIX%\Library                         # [win64]
```

File: recipes/nvmath-python/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        - {{ pin_subpackage("nvmath-python", exact=True) }}
+        - cupy  # TODO: remove me in the next release
+        - numba 0.59.*
+        - cuda-python
+        - cuda-nvcc
+        - cuda-nvrtc
+        - cuda-cudart-dev
+        - cuda-cccl
+        - mathdx 24.04.0.*
+        - cuda-version >=12.0.*,<12.4.a0
```
Comment (jakirkham) [inline recipes/nvmath-python/meta.yaml:123]:
```suggestion
        - cuda-version >=12.0,<12.4.a0
```

File: recipes/nvmath-python/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        - python
+        - {{ pin_subpackage("nvmath-python", exact=True) }}
+        - cupy  # TODO: remove me in the next release
+        - numba 0.59.*
+        - cuda-python
+        - cuda-nvcc
+        - cuda-nvrtc
+        - cuda-cudart-dev
+        - cuda-cccl
+        - mathdx 24.04.0.*
```
Comment (jakirkham) [inline recipes/nvmath-python/meta.yaml:122]:
```suggestion
        - mathdx 24.04.0
```

File: recipes/nvmath-python/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        # On conda-forge using the CUDA 12 packages is more lightweight.
+        - cuda-version 12.*                      # [cuda_compiler_version == "None"]
+        - cuda-cudart-dev                        # [cuda_compiler_version == "None"]
+        - cuda-crt                               # [cuda_compiler_version == "None"]
+        - cuda-version {{ cuda_major }}.*        # [cuda_compiler_version != "None"]
+      run:
+        - python
+        - numpy >=1.23,<2  # TODO: relax me in the next release
+        - pywin32  # [win64]
+        - cuda-version >={{ cuda_major }}.0.*,<{{ cuda_major+1 }}.0.a0  # [cuda_compiler_version != "None"]
```
Comment (jakirkham) [inline recipes/nvmath-python/meta.yaml:78]:
```suggestion
        - cuda-version >={{ cuda_major }}.0.*,<{{ cuda_major+1 }}.0.a0  # [cuda_compiler_version != "None"]
        - cupy-core
```

File: recipes/nvmath-python/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        {% elif cuda_major == 11 %}
+        - cupy  # TODO: remove me in the next release
+        - cudatoolkit
+        {% endif %}
+      {% if cuda_major == 0 %}
+        - cupy-core  # TODO: remove me in the next release
+      run_constrained:
+        # again, nvmath-python is "CUDA agnostic"
+        - cuda-version >=11.0.*,<13.a0
+      {% endif %}
```
Comment (jakirkham) [inline recipes/nvmath-python/meta.yaml:95]:
```suggestion
      run_constrained:
        - cuda-version >=11.0.*,<13.a0. # [cuda_compiler_version == "None"]
```

**Comment by leofang** [comment]
Pair-reviewed and discussed with @jakirkham offline. To deal with the sporadic, stochastic CI errors, we decided to shrink the support matrix of nvmath-python, thereby reducing the amount of Jinja variables, in the staged-recipes PR. Once the feedstock is created, we'll resume the full support matrix (since the feedstock CIs are more reliable).

**Comment by leofang** [comment]
@conda-forge/help-python-c @conda-forge/help-c-cpp the CI is green now 🙂 See also https://github.com/conda-forge/staged-recipes/pull/26820#issuecomment-2201223982

**Comment by leofang** [comment]
@conda-forge/help-python-c @conda-forge/help-c-cpp friendly nudge :)

**Pull request merged**